projecttacoma/bulk-export-app
# Implement general _typeFilter functionality

# Summary

Create ability to add type filters manually by typing them in a text box.

## New behavior

On the query builder page, the tab for type filters now has an input field and a table that shows the active type filters. When you input a filer in the input box and then press the button that says 'add filter' it should appear on the right in the table. In the able, you can remove filters with the 'x', copy them with the copy button, which adds the filter text to the input box so you can create a similar filter for ease of use. You can also have multiple filters created and then only select a subset of those to run the query with. You can do this by checking the filters that you want to include in the query.

On the home page, I updated the input placeholder so that it will not be undefined:undefined when not running locally

## Code changes

- `src/app/providers.tsx` - added some styling to cards that are reused across multiple pages.
- `src/app/query-builder/page.tsx`  - changed the paper to the card for simplicity
- `src/components/export-execution/polling-logs.tsx` - removed the color theme
- `src/components/parameter-pages/type-filter-params-page.tsx` new file for the logic of the type filter table to implement the functionality described above.
- `src/components/parameter-pages/type-filter-table.tsx` new file to implement the functionality to manually create a type filter from a textbox entry.
- `src/components/query-selector/export-type.tsx` - changed the paper to a card.
- `src/components/query-string.tsx` - add new functionality to add type filter parameters to the function call of the export request builder.
- `src/state/type-filter-params-state.ts` - new recoil state to manage the created type filters. Each type filter has a filter value, which is a string representing the type filter and an active value, which is a boolean to check if the filter is checked in the checkbox on the type filter table.
- `src/util/exportRequestBuilder.ts` -  add new functionality to build the export request with type filters.

# Testing guidance

- `npm run check`
- `npm run build`
- `npm run start`
- Navigate through the pages as usual and go to the type filter tab on the query builder page
- Write some type filters like `Coverage?status=active` or `Patient?gender=female` for example
- Make sure that the export request correctly filters using the active type filters
- Check and uncheck some of the filters and make sure that it correctly shows in the export request string at the top of the page.
- Remove filters, and check that it removes from the table and from the export request string
- Click the copy button and check that the input field is populated with the type filter that you would like to copy.
- Add some filters and click remove all. Verify they are deleted the same as above
- Check all/ uncheck all filters and verify that works correctly.
